File: Examples/Analysis_C/Analysis_4C.m
% Multivariate Bayesian Inversion for Classification and Regression
% Analysis 4: comparison of methods (MATLAB script)
% 
% Author: Joram Soch, OvGU Magdeburg
% E-Mail: joram.soch@ovgu.de
% 
% Version History:
% - 21/02/2022, 01:45: first data analysis
% - 28/02/2025, 18:06: final figure display
% - 26/02/2026, 17:13: rewrote for multiple analyses
% - 26/02/2026, 17:28: added MBR, GNB, MLR, SVR, RFR, NNR


clear
close all

%%% Step 1: load data %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% load data
load ../Analysis_4/PAC_specify.mat
load ../Analysis_4/PAC_specify_test_age.mat

% assemble data (MBR)
n1  = numel(sID1);              % number of data points
n2  = numel(sID2);
V1  = eye(n1);                  % observation covariances
V2  = eye(n2);
YA1 = [GM1, WM1];               % data matrices
YA2 = [GM2, WM2];
x1  = y1;                       % label vectors
x2  = y2;
X1  = [x1, ones(size(x1))];     % design matrices
X2  = [x2, ones(size(x2))];
XA1 = c1(:,2:end);              % covariate matrices
XA2 = c2(:,2:end);

% assemble data (SVR)
YB1 = [GM1, WM1, c1];           % feature matrices
YB2 = [GM2, WM2, c2];


%%% Step 2: analyze data %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% specify analysis parameters
meth    = {'MBR', 'GNB', 'MLR', 'SVR', 'RFR', 'NNR'};
prior.x = [0:1:100];            % MBR prior distribution
prior.p = (1/range(prior.x))*ones(size(prior.x));
Dgnb    = 'mvn';                % GNB distribution name
Mmlr    = 'WLS';                % MLR estimation method
Csvm    = 1;                    % SVM cost parameter
Mrf     = 'Bag';                % RF aggregation method
Lnn     = [size(YB1,2)];        % NN hidden layer sizes
Ann     = 'none';               % NN activation function

% preallocate results
M   = numel(meth);
xp  = zeros(n2,M);
r   = zeros( 1,M);
MAE = zeros( 1,M);

% evaluate all methods
for h = 1:M
    
    % Method: multivariate Bayesian regression (MBI)
    % https://github.com/JoramSoch/MBI/blob/main/MATLAB/ML_MBI.m
    if strcmp(meth{h}, 'MBR')
        mba1 = mbitrain(YA1, x1, XA1, V1, 'MBR');
        pp2  = mbitest(YA2, x2, XA2, V2, mba1, prior);
        [m,j]= max(pp2, [], 2);
        xp2  = prior.x(j);
    end;
    
    % Method: Gaussian naive Bayes (custom)
    % https://en.wikipedia.org/wiki/Naive_Bayes_classifier#Probabilistic_model
    if strcmp(meth{h}, 'GNB')
        B_est1  = (X1'*X1)^-1 * X1'*YA1;
        s2_est1 = (1/n1) * sum((YA1-X1*B_est1).^2);
        Si_est1 = diag(s2_est1);
        log_PP  = zeros(n2,numel(prior.x));
        pp2     = zeros(n2,numel(prior.x));
        for i = 1:n2
            y2i = YA2(i,:);
            for j = 1:numel(prior.x)
                x2ij        = [prior.x(j), 1];
                log_PP(i,j) = log(mvnpdf(y2i, x2ij*B_est1, Si_est1)) + ...
                              log(prior.p(j));
            end;
            pp2(i,:) = exp(log_PP(i,:) - mean(log_PP(i,:)));
            pp2(i,:) = pp2(i,:)./trapz(prior.x, pp2(i,:));
        end;
        [m,j]= max(pp2, [], 2);
        xp2  = prior.x(j);
    end;
    
    % Method: multiple linear regression (MACS)
    % https://github.com/JoramSoch/MACS/blob/master/ME_GLM.m
    if strcmp(meth{h}, 'MLR')
        P1  = inv(V1);
        if strcmp(Mmlr, 'OLS')
            b_est1 = (YB1'*YB1)^-1 * YB1'*x1;
        elseif strcmp(Mmlr, 'WLS')
            b_est1 = (YB1'*P1*YB1)^-1 * YB1'*P1*x1;
        else
            b_est1 = zeros(size(YB1,2),1);
        end;
        xp2 = YB2 * b_est1;
    end;
    
    % Method: support vector regression (LibSVM)
    % https://github.com/JoramSoch/ML4ML/blob/main/ML_SVR.m
    if strcmp(meth{h}, 'SVR')
        opt  = sprintf('-s 4 -t 0 -c %s -q', num2str(Csvm));
        svm1 = svmtrain(x1, YB1, opt);
        xp2  = svmpredict(x2, YB2, svm1, '-q');
    end;

    % Method: random forrest regression (MATLAB)
    % https://de.mathworks.com/help/stats/select-predictors-for-random-forests.html
    if strcmp(meth{h}, 'RFR')
        rfe1 = fitrensemble(YB1, x1, 'Method', Mrf);
        xp2  = predict(rfe1, YB2);
    end;

    % Method: neural network regression (MATLAB)
    % https://de.mathworks.com/help/stats/classificationneuralnetwork.html
    if strcmp(meth{h}, 'NNR')
        nn1 = fitrnet(YB1, x1, 'LayerSizes', Lnn, 'Activations', Ann);
        xp2 = predict(nn1, YB2);
    end;
    
    % calculate performance
    xp(:,h) = xp2;
    r(h)    = corr(xp(:,h), x2);
    MAE(h)  = mean(abs(xp(:,h)-x2));
    
end;
fprintf('\n\n');
clear mba1 B_est1 s2_est Si_est1 b_est1 svm1 rfe1 nn1


%%% Step 3: visualize results %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% open figure
figure('Name', 'Analysis 4 (Comparison)', 'Color', [1 1 1], 'Position', [50 50 800 900]);
cols  = 'bcmryw';
x_off = 2;
y_off = 0.95;
hold on;

% plot performance
for h = 1:M
    bar(h, r(h), 0.7, cols(h));
end;
plot([(1-1), (M+2)], [0, 0], ':k', 'LineWidth', 2);
plot([x_off, x_off]+1/2, [0, 1], '-k', 'LineWidth', 1);
axis([(1-1), (M+2), 0, 1]);
set(gca,'Box','On');
set(gca,'XTick',[1:M],'XTickLabel',meth);
legend([meth, {'chance'}], 'Location', 'NorthEast');
xlabel('regression approach', 'FontSize', 16);
ylabel('predictive correlation', 'FontSize', 16);
title('Analysis 4: Comparison', 'FontSize', 24);
text(x_off+1/2, y_off, sprintf('generative   \nmethods   '), ...
     'FontSize', 16, 'HorizontalAlignment', 'right', 'VerticalAlignment', 'middle');
text(x_off+1/2, y_off, sprintf('   discriminative\n   methods'), ...
     'FontSize', 16, 'HorizontalAlignment', 'left', 'VerticalAlignment', 'middle');

File: Examples/Analysis_C/Analysis_4C.py
# -*- coding: utf-8 -*-
"""
Multivariate Bayesian Inversion for Classification and Regression
Analysis 4: comparison of methods (Python script)

Author: Joram Soch, OvGU Magdeburg
E-Mail: joram.soch@ovgu.de

Version History:
- 2022-02-21, 20:29: first version in MATLAB
- 2025-03-05, 18:45: final version in Python
- 2026-06-22, 15:08: rewrote MBR and SVR,
                     implemented RGA measure
- 2026-06-22, 15:30: added GNB, LinReg, RFR, NNR                     
"""


# specify MBI path
MBI_dir = '../../Python/'

# import packages
import os
import numpy as np
import scipy as sp
import sklearn as skl
import matplotlib.pyplot as plt

# import MBI module
orig_dir = os.getcwd()
os.chdir(MBI_dir)
import MBI
os.chdir(orig_dir)


### Step 1: load data #########################################################

# load data
train = sp.io.loadmat('../Analysis_4/PAC_specify.mat')
test  = sp.io.loadmat('../Analysis_4/PAC_specify_test_age.mat')

# assemble data (MBR)
n1  = train['sID1'].size                    # number of data points
n2  = train['sID2'].size
V1  = np.eye(n1)                            # observation covariances
V2  = np.eye(n2)
YA1 = np.c_[train['GM1'], train['WM1']]     # data matrices
YA2 = np.c_[train['GM2'], train['WM2']]
x1  = np.squeeze(train['y1'])               # label vectors
x2  = np.squeeze(test['y2'])
X1  = np.c_[x1, np.ones((n1,1))]            # design matrices
X2  = np.c_[x2, np.ones((n2,1))]
XA1 = train['c1'][:,1:]                     # covariate matrices
XA2 = train['c2'][:,1:]

# assemble data (SVR)                       # feature matrices
YB1 = np.c_[train['GM1'], train['WM1'], train['c1']]
YB2 = np.c_[train['GM2'], train['WM2'], train['c2']]
del train, test

# prepare histograms
x_min = 0
x_max = 100
dx    = 2.5
nx    = int((x_max-x_min)/dx)+1
xb    = np.linspace(x_min, x_max, nx)


### Step 2: analyze data ######################################################

# specify analysis parameters
meth    = ['GNB', 'MBR', 'LinReg', 'SVR', 'RFR', 'NNR']
prior   = {}
prior['x'] = np.arange(0, 100+1, 1)         # MBR prior distribution
prior['p'] = (1/(np.max(prior['x'])-np.min(prior['x'])))*np.ones(prior['x'].size)
Dgnb    = 'mvn'                             # GNB distribution name
Mlinreg = 'WLS'                             # LinReg estimation method
Ksvm    = 'linear'                          # SVM kernel type
Csvm    =  1                                # SVM cost parameter
Nrf     = 100                               # RF number of trees
Lnn     = (YB1.shape[1],)                   # NN hidden layer sizes
Ann     = 'identity'                        # NN activation function

# preallocate results
M   = len(meth)
xp  = np.zeros((n2,M))
r   = np.zeros(M)
MAE = np.zeros(M)
RGA = np.zeros(M)

# evaluate all methods
for h in range(M):
    
    # Method: multivariate Bayesian regression (MBI)
    # https://github.com/JoramSoch/MBI/blob/main/Python/MBI.py
    if meth[h] == 'MBR':
        mba1 = MBI.model(YA1, x1, V=V1, mb_type='MBR').train()
        pp2  = MBI.model(YA2, x2, V=V2, mb_type='MBR').test(mba1, prior)
        ip2  = np.argmax(pp2, axis=1)
        xp2  = prior['x'][ip2]
    
    # Method: Gaussian naive Bayes (custom)
    # https://en.wikipedia.org/wiki/Naive_Bayes_classifier#Probabilistic_model
    if meth[h] == 'GNB':
        B_est1  = np.linalg.inv(X1.T @ X1) @ X1.T @ YA1
        s2_est1 = (1/n1) * np.sum(np.power(YA1 - X1 @ B_est1, 2), axis=0)
        Si_est1 = np.diag(s2_est1)
        log_PP  = np.zeros((n2, prior['x'].size))
        pp2     = np.zeros((n2, prior['x'].size))
        for i in range(n2):
            y2i = np.array([ YA2[i,:] ])
            for j in range(prior['x'].size):
                x2ij        = np.array([ [prior['x'][j], 1] ])
                ptd         = sp.stats.multivariate_normal(
                                  (x2ij @ B_est1)[0,:], Si_est1)
                log_PP[i,j] = np.log(ptd.pdf(y2i))
            pp2[i,:] = np.exp(log_PP[i,:] - np.mean(log_PP[i,:]))
            pp2[i,:] = pp2[i,:] / np.trapz(pp2[i,:], prior['x'])
        ip2     = np.argmax(pp2, axis=1)
        xp2     = prior['x'][ip2]
    
    # Method: linear regression (Python)
    # https://github.com/JoramSoch/MBI/blob/main/Python/cvBMS.py
    if meth[h] == 'LinReg':
        P1  = np.linalg.inv(V1)
        if Mlinreg == 'OLS':
            b_est1 = np.linalg.inv(YB1.T @ YB1) @ YB1.T @ x1
        elif Mlinreg == 'WLS':
            b_est1 = np.linalg.inv(YB1.T @ P1 @ YB1) @ YB1.T @ P1 @ x1
        else:
            b_est1 = np.zeros((YB1.shape[1],1))
        xp2 = YB2 @ b_est1
    
    # Method: support vector regression (scikit-learn)
    # https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVR.html
    if meth[h] == 'SVR':
        svm1 = skl.svm.SVR(kernel=Ksvm, C=Csvm).fit(YB1, x1)
        xp2  = svm1.predict(YB2)
    
    # Method: random forrest regression (scikit-learn)
    # https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestRegressor.html
    if meth[h] == 'RFR':
        rf1 = skl.ensemble.RandomForestRegressor(
            n_estimators=Nrf).fit(YB1, x1)
        xp2 = rf1.predict(YB2)
    
    # Method: neural network regression (scikit-learn)
    # https://scikit-learn.org/stable/modules/generated/sklearn.neural_network.MLPRegressor.html
    if meth[h] == 'NNR':
        nn1 = skl.neural_network.MLPRegressor(
            hidden_layer_sizes=Lnn, activation=Ann).fit(YB1, x1)
        xp2 = nn1.predict(YB2)
    
    # store test set predictions
    xp[:,h] = xp2
    r[h]    = np.corrcoef(xp[:,h], x2)[0,1]
    MAE[h]  = np.mean(np.absolute(xp[:,h]-x2))
    RGA[h]  = MBI.calc_RGA(x2, xp[:,h], 'RGA')

# delete analysis variables
del mba1, pp2, ip2, B_est1, s2_est1, Si_est1, log_PP, P1, b_est1, svm1, rf1, nn1


### Step 3: visualize results #################################################

# open figure
fig   = plt.figure(figsize=(8,9))
ax    = fig.add_subplot(111)
x_off = 2
y_off = 0.975
cols  = [[  0,  32,  96], [  0,   0, 255],
         [192,   0,   0], [255,   0,   0], [255, 192,   0], [255, 255,   0]]

# plot performance
for h in range(M):
    col = tuple([rgb/255 for rgb in cols[h]])
    ax.bar(h+1, RGA[h], width=0.7, color=col, edgecolor='k', label=meth[h])
ax.plot([(1-1), (M+2)], [1/2, 1/2],     ':k', linewidth=2, label='chance')
ax.plot([x_off+1/2, x_off+1/2], [0, 1], '-k', linewidth=1)
ax.axis([(1-1), (M+2), 0, 1])
ax.set_xticks([])
ax.legend(loc='lower right', fontsize=12)
ax.set_xlabel('regression approach', fontsize=20)
ax.set_ylabel('rank graduation accuracy', fontsize=20)
ax.set_title('Analysis 4', fontsize=24, fontweight='bold')
ax.tick_params(axis='both', labelsize=12)
ax.text(x_off+1/2, y_off, 'generative methods   ',
        fontsize=14, ha='right', va='center')
ax.text(x_off+1/2, y_off, '   discriminative methods',
        fontsize=14, ha='left', va='center')

# enable tight layout
fig.tight_layout()